lucene/core/src/java/org/apache/lucene/index/DocumentsWriterPerThreadPool.java

```diff
@@ -16,228 +16,173 @@
       }
     }
-    ThreadState threadState = new ThreadState(null);
-    threadState.lock(); // lock so nobody else will get this ThreadState
-    threadStates.add(threadState);
-    return threadState;
-}
-
-  DocumentsWriterPerThread reset(ThreadState threadState) {
-    assert threadState.isHeldByCurrentThread();
-    final DocumentsWriterPerThread dwpt = threadState.dwpt;
-    threadState.reset();
+    DocumentsWriterPerThread dwpt = dwptFactory.get();
+    dwpt.lock(); // lock so nobody else will get this DWPT
+    dwpts.add(dwpt);
     return dwpt;
   }
-  
-  void recycle(DocumentsWriterPerThread dwpt) {
-    // don't recycle DWPT by default
-  }
 
   // TODO: maybe we should try to do load leveling here: we want roughly even numbers
   // of items (docs, deletes, DV updates) to most take advantage of concurrency while flushing
 
-  /** This method is used by DocumentsWriter/FlushControl to obtain a ThreadState to do an indexing operation (add/updateDocument). */
-  ThreadState getAndLock() {
-    ThreadState threadState = null;
+  /** This method is used by DocumentsWriter/FlushControl to obtain a DWPT to do an indexing operation (add/updateDocument). */
+  DocumentsWriterPerThread getAndLock() throws IOException {
     synchronized (this) {
-      if (freeList.isEmpty()) {
-        // ThreadState is already locked before return by this method:
-        return newThreadState();
-      } else {
-        // Important that we are LIFO here! This way if number of concurrent indexing threads was once high, but has now reduced, we only use a
-        // limited number of thread states:
-        threadState = freeList.remove(freeList.size()-1);
-
-        if (threadState.dwpt == null) {
-          // This thread-state is not initialized, e.g. it
-          // was just flushed. See if we can instead find
-          // another free thread state that already has docs
-          // indexed. This way if incoming thread concurrency
-          // has decreased, we don't leave docs
-          // indefinitely buffered, tying up RAM.  This
-          // will instead get those thread states flushed,
-          // freeing up RAM for larger segment flushes:
-          for(int i=0;i<freeList.size();i++) {
-            ThreadState ts = freeList.get(i);
-            if (ts.dwpt != null) {
-              // Use this one instead, and swap it with
-              // the un-initialized one:
-              freeList.set(i, threadState);
-              threadState = ts;
-              break;
-            }
-          }
+      if (closed) {
+        throw new AlreadyClosedException("DWPTPool is already closed");
+      }
+      // Important that we are LIFO here! This way if number of concurrent indexing threads was once high,
+      // but has now reduced, we only use a limited number of DWPTs. This also guarantees that if we have suddenly
+      // a single thread indexing
+      for (int i = freeList.size()-1; i >= 0; i--) {
```

Comment:
Here you could use the `descendingIterator()` and iterate while just calling remove() on the iterator.

Reply:
👍